app/master/build.py

```diff
@@ -564,6 +569,121 @@ def _generate_unique_symlink_path_for_build_repo(self):
+        q_build.artifacts_zip_file = self._artifacts_zip_file
+        q_build.error_message = self._error_message
+        q_build.postbuild_tasks_are_finished = self._postbuild_tasks_are_finished
+        q_build.setup_failures = self.setup_failures
+        q_build.timing_file_path = self._timing_file_path
+
+        build_artifact_dir = None
+        if self._build_artifact is not None:
+            build_artifact_dir = self._build_artifact.build_artifact_dir
+
+        q_build.build_artifact_dir = build_artifact_dir
+
+        if self._build_artifact is not None:
+            # Clear all old directories
+            session.query(FailedArtifactDirectoriesSchema) \
+                .filter(FailedArtifactDirectoriesSchema.build_id == self._build_id) \
+                .delete()
+
+            # Commit changes so we don't delete the newly added rows later
+            session.commit()
+
+            # Add all the updated versions of the directories
+            for directory in self._build_artifact._get_failed_artifact_directories():  # pylint: disable=protected-access
+                failed_artifact_directory = FailedArtifactDirectoriesSchema(
+                    build_id=self._build_id,
+                    failed_artifact_directory=directory
+                )
+                session.add(failed_artifact_directory)
+
+        if self._build_artifact is not None:
+            # Clear all old directories
+            session.query(FailedSubjobAtomPairsSchema) \
+                .filter(FailedSubjobAtomPairsSchema.build_id == self._build_id) \
+                .delete()
+
+            # Commit changes so we don't delete the newly added rows later
+            session.commit()
+
+            # Add all the updated versions of the data
+            for subjob_id, atom_id in self._build_artifact.get_failed_subjob_and_atom_ids():
+                failed_subjob_and_atom_ids = FailedSubjobAtomPairsSchema(
+                    build_id=self._build_id,
+                    subjob_id=subjob_id,
+                    atom_id=atom_id
+                )
+                session.add(failed_subjob_and_atom_ids)
+
+        q_build.build_parameters = json.dumps(self._build_request.build_parameters())
+
+        fsm_timestamps = {state.lower(): timestamp for state, timestamp in self._state_machine.transition_timestamps.items()}
+        q_build.state = self._status()
+        q_build.queued_ts = fsm_timestamps['queued']
+        q_build.finished_ts = fsm_timestamps['finished']
+        q_build.prepared_ts = fsm_timestamps['prepared']
+        q_build.preparing_ts = fsm_timestamps['preparing']
+        q_build.error_ts = fsm_timestamps['error']
+        q_build.canceled_ts = fsm_timestamps['canceled']
+        q_build.building_ts = fsm_timestamps['building']
+
+        # Subjobs
+        # Clear all old Subjobs and Atoms
+        session.query(SubjobsSchema) \
+            .filter(SubjobsSchema.build_id == self._build_id) \
+            .delete()
```

Comment:
I did it this way because I wasn't sure if we're guaranteed to _always_ have the same amount of subjobs and/or atoms for a build. Maybe this assumption is wrong -- in which case we should just update the subjobs/atoms columns instead of delete+create.

Is my assumption wrong?